src/com/google/javascript/jscomp/AdvancedCoverageInstrumentationCallback.java

```diff
@@ -0,0 +1,205 @@
+package com.google.javascript.jscomp;
+
+import com.google.common.annotations.GwtIncompatible;
+import com.google.debugging.sourcemap.Base64VLQ;
+import com.google.javascript.rhino.IR;
+import com.google.javascript.rhino.Node;
+import java.io.IOException;
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+
+/**
+ * Instrument advanced coverage for javascript.
+ */
+@GwtIncompatible("AdvancedFileInstrumentationData")
+public class AdvancedCoverageInstrumentationCallback extends
+    NodeTraversal.AbstractPostOrderCallback {
+
+  private static final String INSTRUMENT_CODE_FUNCTION_NAME = "instrumentCode";
+  private static final String INSTRUMENT_CODE_FILE_NAME = "InstrumentCode.js";
+  private final AbstractCompiler compiler;
+  private final ParameterMapping parameterMapping;
+  private final Map<String, AdvancedFileInstrumentationData> instrumentationData;
+  private String functionName = "Anonymous";
+
+  public AdvancedCoverageInstrumentationCallback(
+      AbstractCompiler compiler, Map<String, AdvancedFileInstrumentationData> instrumentationData) {
+    this.compiler = compiler;
+    this.instrumentationData = instrumentationData;
+    this.parameterMapping = new ParameterMapping();
+  }
+
+  @Override
+  public void visit(NodeTraversal traversal, Node node, Node parent) {
+    String fileName = traversal.getSourceName();
+    String sourceFileName = node.getSourceFileName();
+
+    // If origin of node is not from sourceFile, do not instrument. This typically occurs when
+    // polyfill code is injected into the sourceFile AST and this check avoids instrumenting it. We
+    // avoid instrumentation as this callback does not distinguish between sourceFile code and
+    // injected code and can result in an error.
+    if (!Objects.equals(fileName, sourceFileName)) {
+      return;
+    }
+
+    // If Source File is base.js or INSTRUMENT_CODE_FILE_NAME, do not instrument as the instrument
+    // function has not yet been defined when base.js will be executing and the implementation file
+    // can not call the Code instrumentation function on itself
+    if (sourceFileName.endsWith("base.js") || sourceFileName.endsWith(INSTRUMENT_CODE_FILE_NAME)) {
```

Comment:
Let's not hardcode "base.js" here. Could we instead have a check like "if this is the first source file in the program, do not instrument"?

Reply:
Changed as discussed.